### PR TITLE
Deduplication improvements

### PR DESCRIPTION
**Pull request by acheronw**
Wrote a script that deduplicates (using the minhashes) against every earlier batch (treating directory names as dates).

### CODE REVIEW

File: scripts/lsh.py [hunk 1]
```diff
@@ -309,6 +321,68 @@ def other_main(args):
...
+
+    # We suppose here that the final part of the input directory is a
+    # date-like string e.g.: 06_filtered/2022_12/
+    input_date = os.path.basename(os.path.normpath(args.input_dir))
+    past_batches = collect_previous_dirs(args.cumulative_dir, input_date)
+    number_of_past_batches = len(past_batches)
+
+    with TemporaryDirectory() as tmp_root_dir:
+        current_input_dir = args.input_dir
+        i = 0
```
Comment (DavidNemeskey) [inline scripts/lsh.py:363]: Just... [don't](https://docs.python.org/3/library/functions.html#enumerate).

File: scripts/lsh.py [hunk 2]
```diff
@@ -309,6 +321,68 @@ def other_main(args):
...
+                # final output dir
+                current_output_dir = args.output_dir
+            else:
+                past_batch_date = os.path.basename(past_batch)
+                # There are still cross-deduplications to do, the results will go
+                # to the tmp output dir.
+                current_output_dir = f"{tmp_root_dir}/{input_date}" \
+                                     f"_against_{past_batch_date}/"
+            logging.info(f"Cross-deduplicating {current_input_dir} with {past_batch}, "
+                  f"moving results to {current_output_dir}")
```
Comment (DavidNemeskey) [inline scripts/lsh.py:377]: Please line up the continuation line with the previous one. Also, in general I am not against `"` being used as the string quotation mark, but in this codebase we use `'` instead.

File: scripts/lsh.py [hunk 3]
```diff
@@ -309,6 +321,68 @@ def other_main(args):
...
+            pairwise_directory_deduplication(current_input_dir, current_output_dir,
+                                             past_batch, args.processes,
+                                             args.permutations, args.threshold)
+            current_input_dir = current_output_dir
+    # TODO do we need any extra statistics to log at the end?
```
Comment (DavidNemeskey) [inline scripts/lsh.py:382]: Probably not, but you'll need to delete the temporary directory. Also, three empty lines are one too much (again, pyflakes).

Comment (DavidNemeskey) [inline scripts/lsh.py:382]: Probably not, but you'll need to delete the temporary directory. Also, three empty lines are one too much (again, pyflakes).

File: scripts/lsh.py [hunk 4]
```diff
@@ -62,6 +64,15 @@ def parse_arguments():
                                    'values for the corpus to cross-deduplicate '
                                    'with.')
 
+    parser_cumulative_cross = subparsers.add_parser(
+        'cumulative', help ='Remove all documents from a corpus that are found in any of the earlier corpora'
+    )
+    parser_cumulative_cross.set_defaults(command="cumulative")
+    parser_cumulative_cross.add_argument('--cumulative-dir', '-c', required=True,
+                                         help='the directory which contains the subdirectories '
+                                            'with the minhash values of the corpora to '
```
Comment (DavidNemeskey) [inline scripts/lsh.py:73]: Please line the strings up.

File: scripts/lsh.py [hunk 5]
```diff
@@ -62,6 +64,15 @@ def parse_arguments():
                                    'values for the corpus to cross-deduplicate '
                                    'with.')
 
+    parser_cumulative_cross = subparsers.add_parser(
+        'cumulative', help ='Remove all documents from a corpus that are found in any of the earlier corpora'
```
Comment (DavidNemeskey) [inline scripts/lsh.py:68]: This line is way too long, and it should be split.

File: scripts/lsh.py [hunk 6]
```diff
@@ -62,6 +64,15 @@ def parse_arguments():
...
+    )
+    parser_cumulative_cross.set_defaults(command="cumulative")
+    parser_cumulative_cross.add_argument('--cumulative-dir', '-c', required=True,
+                                         help='the directory which contains the subdirectories '
```
Comment (DavidNemeskey) [inline scripts/lsh.py:72]: These lines are too long as well. Use the following indentation:

```Python
long_function_call(
    arg1, arg2,
    arg3
)
```

**Review by DavidNemeskey — CHANGES_REQUESTED** [review]
I added a few comments. :)

File: scripts/lsh.py [hunk 7]
```diff
@@ -319,7 +393,16 @@ def main():
     install_mp_handler()
     os.nice(20)
 
-    self_main(args) if args.command == 'self' else other_main(args)
+    if args.command == "self":
+        single_directory_deduplication(args)
+    elif args.command == "other":
+        pairwise_directory_deduplication(args.input_dir, args.output_dir,
+                                         args.cross_dir, args.processes,
+                                         args.permutations, args.threshold)
```
[the remark this answers is not in the document]
Comment (acheronw) [reply scripts/lsh.py:401]: I was already considering doing that.

File: scripts/lsh.py [hunk 8]
```diff
@@ -309,6 +321,68 @@ def other_main(args):
...
+def collect_previous_dirs(path, deadline_date):
+    """ Collects the directories which are directly under the path given
+     and whose name, when interpreted as a date, are earlier than the
+     deadline_date
+
+     TODO If we want to use this functionality in other scripts, it should be
+      moved to utils.py."""
+
+    logging.info(f"We are looking for dirs older than {deadline_date} in {path}")
+    collected_dirs= []
```
[the remark this answers is not in the document]
Comment (acheronw) [reply scripts/lsh.py:333]: You mean the missing space before the = operator?

File: scripts/lsh.py (hunk 3 above)
Comment (acheronw) [reply scripts/lsh.py:382]: I opened the temp dir with:
with TemporaryDirectory() as tmp_root_dir:
and it is supposed to automatically delete it when the "with" block ends. And it did so when I tested it.

File: scripts/lsh.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DavidNemeskey) [reply scripts/lsh.py]: Yes.

File: scripts/autonomous_cross_deduplicator.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+from argparse import ArgumentParser
+import logging
+from pathlib import Path
+import re
+
+from lsh import cumulative_directory_deduplication
+
+def parse_arguments():
+    parser = ArgumentParser(description=__doc__)
+    parser.add_argument('--log-level', '-L', type=str, default='info',
```
Comment (DavidNemeskey) [inline scripts/autonomous_cross_deduplicator.py:18]: Please put this at the end of the list (after `-P`), after the task-related arguments.

File: scripts/autonomous_cross_deduplicator.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                        help='the number of permutations per paragraph (256).')
+    parser.add_argument('--processes', '-P', type=int, default=1,
+                        help='number of worker processes to use (max is the '
+                             'num of cores, default: 1). Note that in order '
+                             'to deduplicate documents, much memory might be '
+                             'needed, so it is a good idea to be conservative '
+                             'with the number of processes.')
+    parser.add_argument('--threshold', '-t', type=float, default=0.9,
+                        help='the Jaccard similarity threshold (0.9).')
+
```
Comment (DavidNemeskey) [inline scripts/autonomous_cross_deduplicator.py:32]: No need for an empty line.

File: scripts/autonomous_cross_deduplicator.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    parser.add_argument('--working-dir', '-w', required=True,
+                        help='the directory that contains the batches')
+    parser.add_argument('--processed-dir', type=str, default='07c_minhash_full',
+                        help='the directory containing the fully processed '
+                             'minhashes (07c_minhash_full).')
+    parser.add_argument('--inputs-dir', type=str, default='07b_minhash_self',
+                        help='the directory containing the self-deduplicated '
+                             'minhashes (07b_minhash_self).')
+
```
Comment (DavidNemeskey) [inline scripts/autonomous_cross_deduplicator.py:41]: Too many empty lines (pyflakes should complain about this).

File: scripts/autonomous_cross_deduplicator.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    parser.add_argument('--processes', '-P', type=int, default=1,
+                        help='number of worker processes to use (max is the '
+                             'num of cores, default: 1). Note that in order '
+                             'to deduplicate documents, much memory might be '
+                             'needed, so it is a good idea to be conservative '
+                             'with the number of processes.')
+    parser.add_argument('--threshold', '-t', type=float, default=0.9,
+                        help='the Jaccard similarity threshold (0.9).')
+
+    parser.add_argument('--working-dir', '-w', required=True,
```
Comment (DavidNemeskey) [inline scripts/autonomous_cross_deduplicator.py:33]: `type=Path`

File: scripts/autonomous_cross_deduplicator.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                        help='the directory containing the fully processed '
+                             'minhashes (07c_minhash_full).')
+    parser.add_argument('--inputs-dir', type=str, default='07b_minhash_self',
+                        help='the directory containing the self-deduplicated '
+                             'minhashes (07b_minhash_self).')
+
+
+
+    args = parser.parse_args()
+    if not Path(args.working_dir).is_dir():
```
Comment (DavidNemeskey) [inline scripts/autonomous_cross_deduplicator.py:45]: See comment for line 33 and then there is no need for the `Path` conversion.

File: scripts/autonomous_cross_deduplicator.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                             'with the number of processes.')
+    parser.add_argument('--threshold', '-t', type=float, default=0.9,
+                        help='the Jaccard similarity threshold (0.9).')
+
+    parser.add_argument('--working-dir', '-w', required=True,
+                        help='the directory that contains the batches')
+    parser.add_argument('--processed-dir', type=str, default='07c_minhash_full',
+                        help='the directory containing the fully processed '
+                             'minhashes (07c_minhash_full).')
+    parser.add_argument('--inputs-dir', type=str, default='07b_minhash_self',
```
[the remark this answers is not in the document]
Comment (DavidNemeskey) [reply scripts/autonomous_cross_deduplicator.py:38]: Also, `inputs_dir` should be the first argument, `processed_dir` (I would call it _output_, though) second. Also, they should be positional arguments, not be optional.

File: scripts/autonomous_cross_deduplicator.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+def main():
+    args = parse_arguments()
+    working_dir = Path(args.working_dir)
+
+    logging.basicConfig(
+        level=getattr(logging, args.log_level.upper()),
+        format='%(asctime)s - %(process)s - %(levelname)s - %(message)s'
+    )
+
+    processed_dir =working_dir.joinpath(args.processed_dir)
```
Comment (DavidNemeskey) [inline scripts/autonomous_cross_deduplicator.py:92]: pyflakes

**Review by DavidNemeskey — CHANGES_REQUESTED** [review]
A few more comments. :)

**Review by DavidNemeskey — APPROVED** [review]

File: scripts/lsh.py [hunk 16]
```diff
@@ -282,31 +249,92 @@ def self_main(args):
...
+def pairwise_directory_deduplication(input_dir: Path,
+                                     output_dir: Path,
+                                     cross_dir: Path,
+                                     processes: int,
+                                     permutations: int,
+                                     threshold: float):
     """The "real" main function of the "other" mode."""
-    if not os.path.isdir(args.output_dir):
-        os.makedirs(args.output_dir)
+    Path(output_dir).mkdir(exist_ok=True)
```
Comment (DavidNemeskey) [inline scripts/lsh.py:259]: No need for `Path(...)` if it is already one. Also `parents=True`.

**Review by DavidNemeskey — CHANGES_REQUESTED** [review]
Almost there...

File: setup.py [hunk 17]
```diff
@@ -87,7 +88,7 @@ def readme():
           'scripts/reparse_documents.py',
           'scripts/fix_corpus.py',
           'scripts/extract_htmls.py',
-          'scripts/find_new_dumps',
+          'scripts/find_new_dumps.py',
```
Comment (DavidNemeskey) [inline setup.py:91]: :+1:

**Review by DavidNemeskey — APPROVED** [review]

**Pull request merged**